### PR TITLE
[UNR-2907] Merge 4.24.2, update to master, incorporate fixes

### PR DESCRIPTION
**Pull request by mironec**
#### Description
Updates the 4.24 branch with 4.24.2 support, makes it compatible with 4.22 and 4.23 (hopefully). Fixes some random errors and bugs. Incorporates @cmsmithio's fixes for deprecation warnings. 

Engine PR: https://github.com/improbableio/UnrealEngine/pull/354

#### Release note
Not yet.

#### Tests
Ran the example project. Trying to pass CI.

#### Documentation
Left comments in code about things we should revert/change in the future, should probably create tickets.

#### Primary reviewers
@cmsmithio @joshuahuburn

### CODE REVIEW

**Comment by mironec** [comment]
The only real changes are:
https://github.com/spatialos/UnrealGDK/pull/1827/commits/342656afea1a3cbd133951630a2469e664318e63
https://github.com/spatialos/UnrealGDK/pull/1827/commits/2ecc3f0ac797261d2adad5806a18b57522820675
https://github.com/spatialos/UnrealGDK/pull/1827/commits/9bf3ec64053f12fc24fa038a72730a53e356c9f4
https://github.com/spatialos/UnrealGDK/pull/1827/commits/67adbf0af5f214f13131ac72d157310fa0d4fbe3
Sorry for the review burden.

File: SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs [hunk 1]
```diff
@@ -14,15 +14,21 @@ public class SpatialGDK : ModuleRules
...
     {
         PCHUsage = ModuleRules.PCHUsageMode.UseExplicitOrSharedPCHs;
-        bUseUnity = false;
+#pragma warning disable 0618
+        bFasterWithoutUnity = true;             // Deprecated in 4.24, replace with bUseUnity = false; once we drop support for 4.23
+        if (Target.Version.MinorVersion == 24)  // Due to a bug in 4.24, bFasterWithoutUnity is inversed, fixed in master, so should hopefully roll into the next release, remove this once it does
+        {
+            bFasterWithoutUnity = false;
+        }
+#pragma warning restore 0618
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs]: New and disgusting change I had to make. (1)

File: SpatialGDK/Source/SpatialGDKEditor/SpatialGDKEditor.Build.cs [hunk 2]
```diff
@@ -7,7 +7,13 @@ public class SpatialGDKEditor : ModuleRules
...
 	{
 		PCHUsage = PCHUsageMode.UseExplicitOrSharedPCHs;
-        bUseUnity = false;
+#pragma warning disable 0618
+        bFasterWithoutUnity = true;             // Deprecated in 4.24, replace with bUseUnity = false; once we drop support for 4.23
+        if (Target.Version.MinorVersion == 24)  // Due to a bug in 4.24, bFasterWithoutUnity is inversed, fixed in master, so should hopefully roll into the next release, remove this once it does
+        {
+            bFasterWithoutUnity = false;
+        }
+#pragma warning restore 0618
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKEditor/SpatialGDKEditor.Build.cs]: New and disgusting change I had to make. (2)

File: SpatialGDK/Source/SpatialGDKEditorCommandlet/SpatialGDKEditorCommandlet.Build.cs [hunk 3]
```diff
@@ -7,7 +7,13 @@ public class SpatialGDKEditorCommandlet : ModuleRules
...
 	{
 		PCHUsage = PCHUsageMode.UseExplicitOrSharedPCHs;
-        bUseUnity = false;
+#pragma warning disable 0618
+        bFasterWithoutUnity = true;             // Deprecated in 4.24, replace with bUseUnity = false; once we drop support for 4.23
+        if (Target.Version.MinorVersion == 24)  // Due to a bug in 4.24, bFasterWithoutUnity is inversed, fixed in master, so should hopefully roll into the next release, remove this once it does
+        {
+            bFasterWithoutUnity = false;
+        }
+#pragma warning restore 0618
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKEditorCommandlet/SpatialGDKEditorCommandlet.Build.cs]: New and disgusting change I had to make. (3)

File: SpatialGDK/Source/SpatialGDKEditorToolbar/SpatialGDKEditorToolbar.Build.cs [hunk 4]
```diff
@@ -7,7 +7,13 @@ public class SpatialGDKEditorToolbar : ModuleRules
...
     {
         PCHUsage = ModuleRules.PCHUsageMode.UseExplicitOrSharedPCHs;
-        bUseUnity = false;
+#pragma warning disable 0618
+        bFasterWithoutUnity = true;             // Deprecated in 4.24, replace with bUseUnity = false; once we drop support for 4.23
+        if (Target.Version.MinorVersion == 24)  // Due to a bug in 4.24, bFasterWithoutUnity is inversed, fixed in master, so should hopefully roll into the next release, remove this once it does
+        {
+            bFasterWithoutUnity = false;
+        }
+#pragma warning restore 0618
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKEditorToolbar/SpatialGDKEditorToolbar.Build.cs]: New and disgusting change I had to make. (4)

File: SpatialGDK/Source/SpatialGDKServices/SpatialGDKServices.Build.cs [hunk 5]
```diff
@@ -7,7 +7,13 @@ public class SpatialGDKServices : ModuleRules
...
 	{
 		PCHUsage = PCHUsageMode.UseExplicitOrSharedPCHs;
-        bUseUnity = false;
+#pragma warning disable 0618
+        bFasterWithoutUnity = true;             // Deprecated in 4.24, replace with bUseUnity = false; once we drop support for 4.23
+        if (Target.Version.MinorVersion == 24)  // Due to a bug in 4.24, bFasterWithoutUnity is inversed, fixed in master, so should hopefully roll into the next release, remove this once it does
+        {
+            bFasterWithoutUnity = false;
+        }
+#pragma warning restore 0618
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKServices/SpatialGDKServices.Build.cs]: New and disgusting change I had to make. (5)

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDKTests.Build.cs [hunk 6]
```diff
@@ -7,7 +7,13 @@ public class SpatialGDKTests : ModuleRules
...
 	{
 		PCHUsage = PCHUsageMode.UseExplicitOrSharedPCHs;
-        bUseUnity = false;
+#pragma warning disable 0618
+        bFasterWithoutUnity = true;             // Deprecated in 4.24, replace with bUseUnity = false; once we drop support for 4.23
+        if (Target.Version.MinorVersion == 24)  // Due to a bug in 4.24, bFasterWithoutUnity is inversed, fixed in master, so should hopefully roll into the next release, remove this once it does
+        {
+            bFasterWithoutUnity = false;
+        }
+#pragma warning restore 0618
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDKTests.Build.cs]: New and disgusting change I had to make. (6)

File: ci/unreal-engine.version [hunk 7]
```diff
@@ -1,2 +1,3 @@
+HEAD feature/4.24.2-merge
```
Comment (mironec) [inline ci/unreal-engine.version]: Added new version to test against in CI

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp [hunk 8]
```diff
@@ -71,11 +71,19 @@ int32 USpatialNetConnection::IsNetReady(bool Saturate)
...
+void USpatialNetConnection::UpdateLevelVisibility(const struct FUpdateLevelVisibilityLevelInfo& LevelVisibility)
+#endif
 {
 	SCOPE_CYCLE_COUNTER(STAT_SpatialNetConnectionUpdateLevelVisibility);
 
+#if ENGINE_MINOR_VERSION <= 23
 	UNetConnection::UpdateLevelVisibility(PackageName, bIsVisible);
+#else
+	UNetConnection::UpdateLevelVisibility(LevelVisibility);
+#endif
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp]: @cmsmithio this it the change for UpdateLevel Visibility

**Comment by mironec** [comment]
New change to make Schema generation tests pass
https://github.com/spatialos/UnrealGDK/pull/1827/commits/63625e9e54d6ef1391b2ac306ca7ca9e28132f46
This is temporary, until we fix 4.22, 4.23 then master.

File: SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/ReferenceCountedLockingPolicy.cpp [hunk 9]
```diff
@@ -75,12 +75,18 @@ ActorLockToken UReferenceCountedLockingPolicy::AcquireLock(AActor* Actor, FStrin
 	return NextToken++;
 }
 
-void UReferenceCountedLockingPolicy::ReleaseLock(ActorLockToken Token)
+bool UReferenceCountedLockingPolicy::ReleaseLock(const ActorLockToken Token)
```
Comment (jayprobable) [inline SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/ReferenceCountedLockingPolicy.cpp]: curious why this is not const ActorLockToken & ?

**Comment by mironec** [comment]
This is superseded by https://github.com/spatialos/UnrealGDK/pull/1837.

**Pull request closed without merging**